src/storage/src/hummock/store/version.rs

```diff
@@ -310,63 +305,38 @@ impl HummockReadVersion {
+                        staging_sst_ref.imm_ids.iter().copied().collect();
+
+                    let mut intersect_imm_ids: HashSet<u64> = HashSet::default();
+                    let mut check_consecutive = false;
+                    let mut first_pos = 0;
+                    for (idx, data) in self.staging.data.iter().enumerate() {
+                        if let Some(imm) = data.to_imm()
+                            && sst_imm_ids.contains(&imm.batch_id())
+                        {
+                            assert!(intersect_imm_ids.is_empty() || check_consecutive);
+                            if intersect_imm_ids.is_empty() {
+                                first_pos = idx;
+                            }
+                            intersect_imm_ids.insert(imm.batch_id());
+                            check_consecutive = true;
+                        } else {
+                            check_consecutive = false;
+                        }
+                    }
 
                     if !intersect_imm_ids.is_empty() {
                         // Check 2)
-                        debug_assert!(check_subset_preserve_order(
-                            intersect_imm_ids.iter().copied(),
-                            self.staging.imm.iter().map(|imm| imm.batch_id()).rev(),
-                        ));
-
-                        // Check 3) and replace imms with a staging sst
-                        for imm_id in &intersect_imm_ids {
-                            if let Some(imm) = self.staging.imm.back() {
-                                if *imm_id == imm.batch_id() {
-                                    self.staging.imm.pop_back();
-                                }
-                            } else {
-                                let local_imm_ids = self
-                                    .staging
-                                    .imm
-                                    .iter()
-                                    .map(|imm| imm.batch_id())
-                                    .collect_vec();
-
-                                unreachable!(
-                                    "should not reach here staging_sst.size {},
-                                    staging_sst.imm_ids {:?},
-                                    staging_sst.epochs {:?},
-                                    local_imm_ids {:?},
-                                    intersect_imm_ids {:?}",
-                                    staging_sst_ref.imm_size,
-                                    staging_sst_ref.imm_ids,
-                                    staging_sst_ref.epochs,
-                                    local_imm_ids,
-                                    intersect_imm_ids,
-                                );
+                        self.staging
+                            .data
+                            .insert(first_pos, StagingData::Sst(staging_sst_ref));
+                        self.staging.data.retain(|data| match data {
```

Comment:
`retain` will iterate over all imms and ssts. It has been observed that when there are many actors, this cost will be amplified greatly, and cause the event handler stuck. With https://github.com/risingwavelabs/risingwave/pull/16725 merged, in `staging_sst_ref`, we can now know what imm id should be replaced. We shall change to maintain a btreemap on the staging data as suggested in the previous comment.

Reply:
Good catch. I will refactor it later.